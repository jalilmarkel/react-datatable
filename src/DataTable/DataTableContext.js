import React, { PureComponent, createContext } from 'react';
import PropTypes from 'prop-types';
import { defaultProps } from './propTypes';

export const defaultState = {
  columns: [],
  selectedRows: [],
  internalCell: false,
  ...defaultProps,
};

export const DataTableContext = createContext(defaultState);

export class DataTableProvider extends PureComponent {
  static propTypes = {
    initialState: PropTypes.object.isRequired,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node,
    ]).isRequired,
  };

  render() {
    const { children, initialState } = this.props;

    return (
      <DataTableContext.Provider value={initialState}>
        {children}
      </DataTableContext.Provider>
    );
  }
}

export const DataTableConsumer = DataTableContext.Consumer;
