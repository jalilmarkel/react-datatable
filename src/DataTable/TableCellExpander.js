import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components';
import ExpanderButton from './ExpanderButton';

const TableCellExpanderStyle = styled.div`
  display: flex;
  flex: 0 0 42px;
  align-items: center;
  box-sizing: border-box;
  vertical-align: middle;
  white-space: nowrap;
  line-height: normal;
  font-weight: 400;
  font-size: ${props => props.theme.rows.fontSize};
  color: ${props => props.theme.rows.fontColor};
  height: ${props => props.theme.rows.height};
  padding: 0;
`;

class TableCellExpander extends PureComponent {
  static propTypes = {
    column: PropTypes.object,
    row: PropTypes.object,
    index: PropTypes.number,
    type: PropTypes.oneOf(['checkbox', 'cell', 'expander']),
    onToggled: PropTypes.func,
    expanded: PropTypes.bool,
  };

  static defaultProps = {
    column: {},
    row: {},
    index: null,
    type: null,
    onToggled: null,
    expanded: false,
  };

  handleCellClick = e => {
    e.stopPropagation();
  };

  render() {
    const {
      column,
      row,
      type,
      onToggled,
      expanded,
      index,
    } = this.props;

    return (
      <TableCellExpanderStyle
        type={type}
        column={column}
        onClick={this.handleCellClick}
      >
        <ExpanderButton
          onToggled={onToggled}
          data={row}
          expanded={expanded}
          index={index}
        />
      </TableCellExpanderStyle>
    );
  }
}

export default withTheme(TableCellExpander);
