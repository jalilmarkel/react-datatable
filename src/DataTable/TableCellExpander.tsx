import * as React from 'react';
import styled from 'styled-components';
import { CellBase } from './Cell';
import ExpanderButton from './ExpanderButton';
import { ExpandableIcon } from './types';

const TableCellExpanderStyle = styled(CellBase)`
	white-space: nowrap;
	font-weight: 400;
	min-width: 48px;
	${({ theme }) => theme.expanderCell.style};
`;

type TableCellExpanderProps<T> = {
	disabled: boolean;
	expanded: boolean;
	expandableIcon: ExpandableIcon;
	id: string | number;
	row: T;
	onToggled: (row: T) => void;
};

function TableCellExpander<T>({
	row,
	expanded = false,
	expandableIcon,
	id,
	onToggled,
	disabled = false,
}: TableCellExpanderProps<T>): JSX.Element {
	return (
		<TableCellExpanderStyle onClick={(e: React.MouseEvent) => e.stopPropagation()} noPadding>
			<ExpanderButton
				id={id}
				row={row}
				expanded={expanded}
				expandableIcon={expandableIcon}
				disabled={disabled}
				onToggled={onToggled}
			/>
		</TableCellExpanderStyle>
	);
}

export default TableCellExpander;
