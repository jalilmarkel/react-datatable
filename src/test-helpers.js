import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components';
import theme from './themes/default';

export const renderWithTheme = (tree, ...args) =>
  render(<ThemeProvider theme={theme()}>{tree}</ThemeProvider>, ...args);

