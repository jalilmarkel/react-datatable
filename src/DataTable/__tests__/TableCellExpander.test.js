import 'jest-styled-components';
import React from 'react';
import { cleanup } from 'react-testing-library';
import { renderWithTheme } from '../../test-helpers';
import { DataTableProvider, defaultState } from '../DataTableContext';
import TableCellExpander from '../TableCellExpander';

afterEach(cleanup);

test('component <TableCellExpander /> should render correctly', () => {
  const { container } = renderWithTheme(
    <DataTableProvider initialState={{
      ...defaultState,
    }}
    >
      <TableCellExpander row={{}} column={{ selector: 'test' }} />
    </DataTableProvider>,
  );

  expect(container.firstChild).toMatchSnapshot();
});
