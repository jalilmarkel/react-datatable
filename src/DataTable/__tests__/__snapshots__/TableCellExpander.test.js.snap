// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component <TableCellExpander /> should render correctly 1`] = `
.c1 {
  outline: none;
  border: none;
  display: block;
  width: 40px;
  height: 40px;
  background-color: transparent;
  color: rgba(0,0,0,.54);
}

.c1:disabled {
  color: rgba(0,0,0,.12);
}

.c1:hover {
  cursor: pointer;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding-left: calc(48px / 6);
  padding-right: calc(48px / 6);
  padding-top: calc(48px / 12);
  padding-bottom: calc(48px / 12);
  -webkit-flex: 0 0 56px;
  -ms-flex: 0 0 56px;
  flex: 0 0 56px;
  white-space: nowrap;
  font-weight: 400;
  font-size: 13px;
  color: rgba(0,0,0,.87);
  min-height: 48px;
}

.c0:not(:first-child) {
  padding-left: 0;
}

<div
  class="c0"
>
  <button
    class="c1"
    data-testid="expander-button"
  >
    <svg
      fill="currentColor"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
      />
      <path
        d="M0-.25h24v24H0z"
        fill="none"
      />
    </svg>
  </button>
</div>
`;
